File: DataAlgo_forPy/Trees/nodeRef_BiTree.rb
class BinaryTree

  attr_accessor  :key, :leftChild, :rightChild

  def initialize(rootObj = nil, left = nil, right = nil)
    @key = rootObj
    @leftChild = left
    @rightChild = right
  end

  def insertLeft(newNode)
    if self.leftChild.nil?
      self.leftChild = BinaryTree.new(newNode)
    else
      branch = BinaryTree.new(newNode,self.leftChild)
      self.leftChild = branch
    end
  end

  def insertRight(newNode)
    if self.rightChild.nil?
      self.rightChild = BinaryTree.new(newNode)
    else
      branch = BinaryTree.new(newNode,self.rightChild)
      self.rightChild = branch
    end
  end

end

r = BinaryTree.new('a')
r.insertLeft('b')
r.insertRight('c')
r.leftChild.insertLeft('d')
r.leftChild.insertRight('e')
r.rightChild.insertLeft('f')
r.rightChild.insertRight('g')

# Traversal Methods - Preorder, Inorder, Postorder

 #Visit root node, recursive preorder left subtree, recursive preorder right subtree
def preorder(tree)
  if tree
  p tree.key
  preorder(tree.leftChild)
  preorder(tree.rightChild)
  end
end

#Inorder traversal on left subtree, visit root node, recursive inorder traversal right subtree
def inorder(tree)
  if tree
  inorder(tree.leftChild)
  p tree.key
  inorder(tree.rightChild)
  end
end

# Recursive postorder traversal left subtree and right subtree followed by root node
def postorder(tree)
  if tree
  postorder(tree.leftChild)
  postorder(tree.rightChild)
  p tree.key
  end
end

puts "------ PREORDER ------"
preorder(r)
puts "------ ------ ------ ------"

puts "------ INORDER ------"
inorder(r)
puts "------ ------ ------ ------"

puts "------ POSTORDER ------"
postorder(r)
puts "------ ------ ------ ------"

File: DataAlgo_forPy/Trees/nodeRef_BiTree.js
class BinaryTree {
  constructor(rootObj, leftChild = null, rightChild = null) {
    this.key = rootObj
    this.leftChild = leftChild
    this.rightChild = rightChild
  }

  insertLeft(newNode) {
    if (!this.leftChild) {
      this.leftChild = new BinaryTree(newNode)
    } else {
      const node = new BinaryTree(newNode)
      node.leftChild = this.leftChild
      this.leftChild = node
    }
  }

  insertRight(newNode) {
    if (!this.rightChild) {
      this.rightChild = new BinaryTree(newNode)
    } else {
      const node = new BinaryTree(newNode)
      node.rightChild = this.rightChild
      this.rightChild = node
    }
  }

  getRightChild() {
    return this.rightChild
  }

  getLeftChild() {
    return this.leftChild
  }

  setRootVal(obj) {
    this.key = obj
  }

  getRootVal() {
    return this.key
  }

}

const preorder = (tree) => {
  if (tree) {
    console.log(tree.key)
    preorder(tree.leftChild)
    preorder(tree.rightChild)
  }
}

const inorder = (tree) => {
  if (tree) {
    inorder(tree.leftChild)
    console.log(tree.key)
    inorder(tree.rightChild)
  }
}

const postorder = (tree) => {
  if (tree) {
    postorder(tree.rightChild)
    postorder(tree.leftChild)
    console.log(tree.key)
  }
}

const r = new BinaryTree('a')
r.insertLeft('b')
r.insertRight('c')
r.leftChild.insertLeft('d')
r.leftChild.insertRight('e')
r.rightChild.insertLeft('f')
r.rightChild.insertRight('g')

console.log("------ PREORDER ------")
preorder(r)
console.log("------ ------ ------ ------")

console.log("------ INORDER ------")
inorder(r)
console.log("------ ------ ------ ------")

console.log("------ POSTORDER ------")
postorder(r)
console.log("------ ------ ------ ------")
